sdk/Pulumi/Provider/PropertyValueSerializer.cs

```diff
@@ -0,0 +1,747 @@
+            {
+                var enumValue = Enum.GetName(targetType, value);
+                return enumValue == null ? PropertyValue.Null : new PropertyValue(enumValue);
+            }
+
+            if (value is IOutput output)
+            {
+                var data = await output.GetDataAsync().ConfigureAwait(false);
+                if (!data.IsKnown)
+                {
+                    return PropertyValue.Computed;
+                }
+
+                var outputValue = await Serialize(data.Value);
+                var dependantResources = ImmutableArray.CreateBuilder<string>();
+                foreach (var resource in data.Resources)
+                {
+                    var urn = await resource.Urn.GetValueAsync("").ConfigureAwait(false);
+                    dependantResources.Add(urn);
+                }
+
+                var outputProperty = new PropertyValue(new OutputReference(
+                    value: outputValue,
+                    dependencies: dependantResources.ToImmutableArray()));
+
+                if (data.IsSecret)
+                {
+                    return new PropertyValue(outputProperty);
+                }
+
+                return outputProperty;
+            }
+
+            if (value is InputArgs inputArgs)
+            {
+                var inputsMap = await inputArgs.ToDictionaryAsync().ConfigureAwait(false);
+                return await Serialize(inputsMap);
+            }
+
+            if (targetType.IsClass || targetType.IsValueType)
+            {
+                var propertyObject = ImmutableDictionary.CreateBuilder<string, PropertyValue>();
+                foreach (var property in targetType.GetProperties())
+                {
+                    var propertyName = PropertyName(property);
+                    var propertyValue = property.GetValue(value);
+                    var serializedProperty = await Serialize(propertyValue);
+                    propertyObject.Add(propertyName, serializedProperty);
+                }
+
+                return new PropertyValue(propertyObject.ToImmutableDictionary());
+            }
+
+            return PropertyValue.Null;
+        }
+
+        private string DeserializationError(
+            PropertyValueType expected,
+            PropertyValueType actual,
+            Type targetType,
+            string[] path)
+        {
+            if (path.Length == 1 && path[0] == "$")
+            {
+                return $"Error while deserializing value of type {targetType.Name} from property value of type {actual}. "
+                       + $"Expected {expected} instead.";
+            }
+
+            var propertyPath = $"[" + string.Join(", ", path) + "]";
+            return $"Error while deserializing value of type {targetType.Name} from property value of type {actual}. "
+                   + $"Expected {expected} instead at path {propertyPath}.";
+        }
+
+        public T Deserialize<T>(PropertyValue value)
```

Comment:
> I've got a few cases I've been thinking of which I'm not sure if they should be 'errors' or bubbled up as extra values here. Mostly around unknowns and secrets.

You mean returning something like `Result<T, DeserializeError>`?

> But like if you have a property value Secret(Number(4)) and you try and pass that to Deserialize<double> what should we do?

Right now this will fail because `Deserialize<double>` is expecting `Number` but instead it gets `Secret`. A `Secret` number means that the value must be an output/input to allow _secretness_ . Using `Deserialize<Input<double>>` will work in this case. 

This makes me think that, actually `Secret(Number(4))` should not be possible to represent in the first place in the domain model. Instead maybe something like `Output(value=Number(4), secret=True)`?

Reply:
> You mean returning something like Result<T, DeserializeError>?

Probably more like { Value: T?, Secret: bool }, T? because it might have been a nested unknown value. But at this point it's like this is basically just `Output<T>` so maybe Deserialise just has to deserialise into an `Output<T>`?

> This makes me think that, actually Secret(Number(4)) should not be possible to represent in the first place in the domain model. Instead maybe something like Output(value=Number(4), secret=True)?

Well both are valid at the moment, and providers/sdks do use both forms. Also for a provider potentially _any_ input could have been made secret (except for plains?) but then you start looking at something like Ian's Output<T> in provider SDKs to support that.

Also even if we decide that `Secret(Number(4))` is an error to deserialise into just `double` what if the secretness is nested? Like `Array([Secret(Number(1)), Number(2)])` deser'd into `Output<double[]>`. Even though the item is only expected to be a `double` not an output it feels safe to lift the secretness up to the nearest output value?

This is definitely something to think about, and probably get comments from the provider team about how they think this should look based on using the Go provider SDK.

Reply:
>Also for a provider potentially any input could have been made secret (except for plains?)

The way I am thinking about this is that when you `Deserialize<double>` you are explicitly saying: "I am expecting a plain value from the raw wire data format". If the data contains secrets then using `double` wasn't the right choice. Users should then use `Input<double>`


> Deserialise just has to deserialise into an Output<T>?

Users would almost always want `T` from here, not the `Output<T>` when deserializing objects or values. In fact, they could just use `Deserialize<Input<T>>` and it will work for plain values as well, I believe